src/pipecat/services/aws_nova_sonic/aws.py

```diff
@@ -0,0 +1,973 @@
+            pass
+        elif isinstance(frame, UserStartedSpeakingFrame):
+            # print("[pk] UserStartedSpeakingFrame")
+            pass
+        elif isinstance(frame, StopInterruptionFrame):
+            # print("[pk] StopInterruptionFrame")
+            pass
+        elif isinstance(frame, UserStoppedSpeakingFrame):
+            # print("[pk] UserStoppedSpeakingFrame")
+            pass
+
+        await self.push_frame(frame, direction)
+
+    async def _handle_context(self, context: OpenAILLMContext):
+        # TODO: reset connection if needed (if entirely new context object provided, for instance)
+        print(f"[pk] received updated context: {context.get_messages_for_initializing_history()}")
+        if not self._context:
+            # We got our initial context - try to finish connecting
+            self._context = AWSNovaSonicLLMContext.upgrade_to_nova_sonic(
+                context, self._system_instruction
+            )
+            self._context_available = True
+            await self._finish_connecting_if_context_available()
+
+    async def _handle_input_audio_frame(self, frame: InputAudioRawFrame):
+        # Wait until we're done sending the assistant response trigger audio before sending audio
+        # from the user's mic
+        if self._triggering_assistant_response:
+            return
+
+        # TODO: check if _audio_input_paused? what causes that?
+        await self._send_user_audio_event(frame.audio)
+
+    async def _handle_bot_stopped_speaking(self):
+        # Protect against back-to-back BotStoppedSpeaking calls, which I've observed
+        if self._handling_bot_stopped_speaking:
+            return
+        self._handling_bot_stopped_speaking = True
+
+        if self._assistant_is_responding:
+            # Consider the assistant finished with their response (after a short delay, to allow for
+            # any FINAL text block to come in).
+            #
+            # TODO: ideally we could base this solely on the LLM output events, but I couldn't
+            # figure out a reliable way to determine when we've gotten our last FINAL text block
+            # after the LLM is done talking.
+            #
+            # First I looked at stopReason, but it doesn't seem like the last FINAL text block is
+            # reliably marked END_TURN (sometimes the *first* one is, but not the last...bug?)
+            #
+            # Then I considered schemes where we tally or match up SPECULATIVE text blocks with
+            # FINAL text blocks to know how many or which FINAL blocks to expect, but user
+            # interruptions throw a wrench in these schemes: depending on the exact timing of the
+            # interruption, we should or shouldn't expect some FINAL blocks.
+            await asyncio.sleep(0.25)
+            self._assistant_is_responding = False
+            await self._report_assistant_response_ended()
+            self._handling_bot_stopped_speaking = False
```

Comment:
nit: We can remove this one, since we are already setting it outside the if.

Reply:
good catch!

Reply:
fixed